### PR TITLE
Declare security property of OpenAPI in methods

### PR DESCRIPTION
**Pull request by frbuceta**
## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by frbuceta** [comment]
Hi @raymondfeng, can you take a look at this and say what you think?

File: packages/authentication/package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "@loopback/core": "^2.13.1",
     "@loopback/eslint-config": "^10.0.4",
     "@loopback/openapi-spec-builder": "^3.0.3",
+    "@loopback/openapi-v3": "^5.1.2",
```
Comment (raymondfeng) [inline packages/authentication/package.json:42]: Let's use `@loopback/rest` for imports as types in `@loopback/openapi-v3` are re-exported in `@loopback/rest`.

File: packages/authentication/tsconfig.json [hunk 2]
```diff
@@ -16,6 +16,9 @@
     {
       "path": "../openapi-spec-builder/tsconfig.json"
     },
+    {
+      "path": "../openapi-v3/tsconfig.json"
+    },
```
Comment (raymondfeng) [inline packages/authentication/tsconfig.json:21]: Let's revert it.

**Comment by raymondfeng** [comment]
@frbuceta Great start. I would appreciate some examples/tests to show the usage of such decorators.

File: packages/authentication/package.json (hunk 1 above)
Comment (frbuceta) [reply packages/authentication/package.json]: Done

File: packages/authentication/tsconfig.json (hunk 2 above)
Comment (frbuceta) [reply packages/authentication/tsconfig.json]: Done

**Comment by frbuceta** [comment]
> @frbuceta Great start. I would appreciate some examples/tests to show the usage of such decorators.

In the next few days I will present some tests

**Comment by frbuceta** [comment]
I have uploaded some tests to verify that this works

File: packages/openapi-v3/src/__tests__/integration/controller-spec.integration.ts [hunk 3]
```diff
@@ -219,6 +222,49 @@ describe('controller spec', () => {
     });
   });
 
+  it('generates a segurity object', () => {
```
Comment (nflaig) [inline packages/openapi-v3/src/__tests__/integration/controller-spec.integration.ts:225]: @frbuceta typo in security

File: packages/authentication/src/decorators/authenticate.decorator.ts [hunk 4]
```diff
@@ -52,14 +52,26 @@ export function authenticate(
 
     if (method && methodDescriptor) {
       // Method
+      MethodDecoratorFactory.createDecorator<SecurityRequirementObject[]>(
+        OAI3Keys.SECURITY_METHOD_KEY,
+        specs.map(o => ({[o.strategy]: []})),
```
Comment (nflaig) [inline packages/authentication/src/decorators/authenticate.decorator.ts:57]: @frbuceta one problem I see with this approach is that if a strategy does not define a `securityScheme` the openapi spec will be invalid

**Review by nflaig — COMMENTED** [review]
@frbuceta Great idea to automatically add the security to each method decorated with `@authenticate`. The only problem I see is that it might be required to only add this for strategies which add a security scheme, else we end up with an invalid api spec.

File: packages/openapi-v3/src/__tests__/integration/controller-spec.integration.ts (hunk 3 above)
Comment (frbuceta) [reply packages/openapi-v3/src/__tests__/integration/controller-spec.integration.ts:225]: Thanks for the review, I correct it

File: packages/authentication/src/decorators/authenticate.decorator.ts (hunk 4 above)
Comment (frbuceta) [reply packages/authentication/src/decorators/authenticate.decorator.ts:57]: Well seen, I had thought about it when I started to program it but I had forgotten. I'm going to see what solutions I have

Comment (nflaig) [reply packages/authentication/src/decorators/authenticate.decorator.ts:57]: Maybe another approach could be to utilize a spec enhancer and allow authentication strategies to specify the security spec, see [gist](https://gist.github.com/nflaig/d8b6b30d0d328eae11eefe3e4ff50dce).

**Review by mschnee — CHANGES_REQUESTED** [review]
This is a great start and I love that we're building in more of the feature set of the OpenAPI spec.  Two points:

First, you're going to need clarify the scope configuration, as defined in the OpenAPI spec:
```yaml
paths:
  /users:
    get:
      summary: Get a list of users
      security:
        - OAuth2: [read]     # <------
      ...
    post:
      summary: Add a user
      security:
        - OAuth2: [write]    # <------
      ...
```
I think that updates to configuration in `@authorize` could theoretically support this, though I'm not familiar enough with the best way to do that.

And second, a *question*,
Why add a new decorator?  If the purpose behind `@security` is to add the security schemes without creating an authentication strategy required by `@authenticate`, I think I have to push back: defining a path object with a security statement, but actually having no security, is not a pattern I'm comfortable with supporting.

@nflaig made a great point about the strategy needing to be defined.  I would like to make the argument that the strategy *should* and *must* be defined.

**Comment by nflaig** [comment]
@mschnee good points, also I am curious how adding the scopes could be achieved through the `@authorize` this sounds rather difficult because the authentication strategy (e.g. OAuth2) comes from `@authenticate` but maybe it is possible.

I created a PoC which would address all the points you mentioned besides the scopes, see [gist](https://gist.github.com/nflaig/d8b6b30d0d328eae11eefe3e4ff50dce). This would extend the `AuthenticationStrategy` interface and allow strategies to provide a security spec and only if provided the security of the openapi spec would be extended which would solve the problem I noted above.

**Pull request closed without merging**